azkaban-common/src/main/java/azkaban/container/models/AzKubernetesV1PodTemplate.java

```diff
@@ -0,0 +1,140 @@
+import io.kubernetes.client.openapi.models.V1Pod;
+import io.kubernetes.client.openapi.models.V1PodSpec;
+import io.kubernetes.client.openapi.models.V1Volume;
+import io.kubernetes.client.openapi.models.V1VolumeMount;
+import io.kubernetes.client.util.Yaml;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * A singleton class which reads a k8s pod-spec yaml file as template. Items like InitContainers,
+ * Volumes and VolumeMounts for the application-container/flow-container will be extracted from the
+ * POD created from this template file.
+ * <p>
+ * Extracted Items are later merged via {@link azkaban.executor.container.KubernetesContainerizedImpl#mergePodSpec(V1PodSpec,
+ * AzKubernetesV1PodTemplate)} with the pod-spec created using {@link AzKubernetesV1SpecBuilder}
+ * <p>
+ * Merging criterion is such that, if the item in the already created pod-spec has the same name as
+ * of the item extracted from the template, then it will be retained.
+ * <p>
+ * This template design enables declaration of static initContainers, volumes, volumeMounts, etc.
+ * which are not job-types but are useful. For example: init container which will fetch the required
+ * certificates/tokens etc. writes to a volume and then that volume will be mounted to the
+ * application-container/flow-container.
+ */
+public class AzKubernetesV1PodTemplate {
+
+  private V1Pod podFromTemplate;
+  private static AzKubernetesV1PodTemplate instance;
+
+  /**
+   * Private constructor to make this class singleton
+   *
+   * @param templatePath th where the template file is located.
+   * @throws IOException If unable to read the template file.
+   */
+  private AzKubernetesV1PodTemplate(String templatePath) throws IOException {
+    File templateFile = Paths.get(templatePath).toFile();
+    this.podFromTemplate = (V1Pod) Yaml.load(templateFile);
+  }
+
+  /**
+   * @param templatePath Path where the template file is located.
+   * @return Singleton instance of this class.
+   * @throws IOException If unable to read the template file.
+   */
+  public static synchronized AzKubernetesV1PodTemplate getInstance(String templatePath)
+      throws IOException {
+    if (null == instance) {
+      instance = new AzKubernetesV1PodTemplate(templatePath);
+    }
+    return instance;
+  }
+
+  /**
+   * @return the {@link V1Pod} POD generated from the template.
+   */
+  public V1Pod getPodFromTemplate() {
+    return this.podFromTemplate;
+  }
+
+  /**
+   * @param filterPredicate Predicate to filter the init containers.
+   * @return The list of filtered init Containers derived from the POD specified in the template.
+   */
+  public List<V1Container> getInitContainers(Predicate<? super V1Container> filterPredicate) {
+    V1PodSpec spec = this.podFromTemplate.getSpec();
```

Comment:
We have the AzKubernetesV1PodTemplate as a singleton. When instantiated, podFromTemplate will always be non-null as we are using "Yaml.load()" which will not return null but will fail in case it is unable to read or even if the file is empty. Hence it will never be null.